gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+            for (const auto & entry : current_tuple) {
+                overloaded{
+                    [&](const BinaryEntry & binary_entry) {
+                        node_visited[binary_entry.var_1] = false;
+                        node_visited[binary_entry.var_2] = false;
+                        adjacent_edges.try_emplace(binary_entry.var_1, vector<SmartEntry>{});
+                        adjacent_edges.try_emplace(binary_entry.var_2, vector<SmartEntry>{});
+                        adjacent_edges[binary_entry.var_1].emplace_back(binary_entry);
+                        adjacent_edges[binary_entry.var_2].emplace_back(binary_entry);
+                    },
+                    [&](const UnaryValueEntry & unary_val_entry) {
+                        node_visited[unary_val_entry.var] = false;
+                        adjacent_edges.try_emplace(unary_val_entry.var, vector<SmartEntry>{});
+                        adjacent_edges[unary_val_entry.var].emplace_back(unary_val_entry);
+                    },
+                    [&](const UnarySetEntry & unary_set_entry) {
+                        node_visited[unary_set_entry.var] = false;
+                        adjacent_edges.try_emplace(unary_set_entry.var, vector<SmartEntry>{});
+                        adjacent_edges[unary_set_entry.var].emplace_back(unary_set_entry);
+                    }}
+                    .visit(entry);
+            }
+
+            SmartTable::Forest forest{};
+            for (auto & var_pair : node_visited) {
+                auto & var = var_pair.first;
+                auto & visited = var_pair.second;
+                if (visited) continue;
+                vector<vector<SmartEntry>> entry_tree;
+                entry_tree.emplace_back();
+                // Recursively build the tree starting from this node
+                build_tree(var, 0, entry_tree, node_visited, adjacent_edges);
+                forest.emplace_back(entry_tree);
+            }
+
+            forests.emplace_back(forest);
+        }
+        return forests;
+    }
+}
+
+// For PB model
+auto make_binary_entry_flag(State & state, Propagators & propagators, const IntegerVariableID & var_1, const IntegerVariableID & var_2, const ConstraintType & c) -> ProofFlag
+{
+    ProofFlag flag, flag_lt, flag_gt; // Is this initialisation okay?
+    WeightedPseudoBooleanTerms var_1_minus_var_2 = {{1_i, var_1}, {-1_i, var_2}};
+    WeightedPseudoBooleanTerms var_2_minus_var_1 = {{1_i, var_2}, {-1_i, var_1}};
+    switch (c) {
+    case EQUAL:
+        // f => var1 == var2
+        flag = propagators.create_proof_flag("bin_eq");
+        propagators.define_pseudoboolean_eq(state, move(var_1_minus_var_2), 0_i, flag);
+
+        // !f => var1 != var2
+        flag_lt = propagators.create_proof_flag("lt");
+        flag_gt = propagators.create_proof_flag("gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag_gt);
```

Comment:
I think this is only working by fluke: `move` can be a destructive operation. You can't rely upon a variable having a sensible value after you move it, but you're using `var_1_minus_var_2` twice.

Reply:
Fixed in c62e8914eb9e01df2a147f01713ff505d7e905e0